docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+* The Cosmos SDK currently lacks a comprehensive ADR for the cryptographic package.
+* Type leakage of specific crypto data types expose backward compatibility and extensibility challenges.
+* The demand for a more flexible and extensible approach to cryptography and address management is high.
+* Architectural changes are necessary to resolve many of the currently open issues.
+* There is a current trend towards modularity in the Interchain stack (e.g. runtime modules)
+* Security implications are a critical consideration during the redesign work.
+
+## Objectives
+
+The key objectives for the CryptoV2 module are:
+
+Modular Design Philosophy
+
+* Establish a flexible and extensible foundation using interfaces to enable the seamless integration of various cryptographic methods and tools.
+
+* Restructure, Refactor, and Decouple: Update the cryptography codebase to ensure modularity and future adaptability.
+
+Documentation & Community Engagement
+
+* Cryptography v2 ADR: Draft a new Architecture Decision Record to guide and document the evolution of the module (this document).
+
+* Enhance documentation to ensure clarity, establish a good practices protocol and promote community engagement, providing a platform for feedback and collaborative growth.
+
+Backward Compatibility & Migration
+
+* Prioritize compatibility with previous module version to avoid disruptions for existing users.
+
+* Design and propose a suitable migration path, ensuring transitions are as seamless as possible.
+
+* Evaluate and decide on the relevance of existing systems and tools, incorporating or deprecating them based on their alignment with the module's new vision.
+
+Developer-Centric Approach
+
+* Prioritize clear, intuitive interfaces and best-practice design principles.
+* Improve Developer Experience: Provide tools, samples, and best practices to foster an efficient and user-friendly development environment.
+
+Leverage Extensibility
+
+* Utilize the module's modular design to support a wide range of cryptographic tools, key types, and methods, ensuring adaptability for future technological advancements.
+* Integrate support for advanced cryptographic features, ensuring the module's position at the forefront of cryptographic technologies.
+
+Quality Assurance
+
+* Enhanced Test Coverage: Improve testing methodologies to ensure the robustness and reliability of the module.
+* Conduct an Audit: After implementation, perform a comprehensive audit to identify potential vulnerabilities and ensure the module's security and stability.
+
+## Technical Goals
+
+As technical goals, the aim is to create a robust, flexible, and future-proof cryptographic module. This is achieved through the following key points:
+
+Wide Hardware Device & Cloud-based HSM Interface Support:
+
+* Design a foundational interface for various hardware devices (Ledger, YubiKey, Thales, etc.) and cloud-based HSMs (Amazon, Azure) to cater to both current and future implementations.
```

Comment:
I think it would help me understand the different components to have examples of how e.g. a YubiKey would fit these interfaces. Is it a `ICryptoProviderBuilder` or a `ISecureStorage` or both? What algorithm does its `Signer` implement? What do its `GetCipher()` and `GetHasher()` methods return?

One end-to-end concrete implementation example (not the actual code, just a description) would go a long way towards clearing up the intention and facilitating feedback.

Reply:
The core idea of the `CryptoProvider` interface is that it abstracts out what 'kind' of entity is behind the iface providing the crypto tools. So, let's take a Ledger HW implementation example (the core idea is the same for any type of backend), I'll be basically following [this](https://github.com/cosmos/cosmos-sdk/blob/ADR/crypto_v2/docs/architecture/adr-071-cryptography-v2.md#introduction-1) diagram:

### `LedgerCryptoProvider` implements `ICryptoProvider`
This is the main API that users (code) will use. From this API, you'll get access to *any* crypto tool that *matches* that backend, this is:
 - `LedgerSigner` implements `ISigner`: this will delegate the entire signing process to the HW, quite similar to what the sdk has today for ledger devices.
 - `LedgerHasher` implements `IHasher`: in this particular device, hashing happens at signing time inside the HW so here this file will implement in code the same hashing logic that the corresponding Ledger does internally.
 - `LedgerVerifier` implements `IVerifier`: here it also makes use of the `LedgerHasher` described above to be able to verify the signed tx.
 - {same idea follows for `LedgerGenerator` and other crypto tools}

So far, we have an API (`LedgerCryptoProvider`) that encapsulates a Ledger device.
Now, could be the case that several LedgerCryptoProvider could exist, each for a different pk. So we need a way to persist AND identify each device to know which implementation to load in memory (in Ledger case could be the serial number for example or a derivation path, just to name a few). The same happens for other devices/providers: they'll need special identifiers and configs (like a cloud-based provider could need a user/pass/url, etc). 

For this, the ADR proposes the creation of another abstraction layer called a `SecureItem`, this iface splits the concepts of WHAT to build, the pk that should be linked to, and WHICH CONFIG should be applied to it at build time, and of course this must be persisted somehow to be reusable. Now, for example let's assume that I want to store all of the above (what to build, id and config) in a json file, in that case I'll create:

### `SecureItemJson` implements `ISecureItem`
Note that this is completely detached from the underlying implementation of the CryptoProvider it points to, this is, I could use the same implementation of `SecureItemJson` to store how to build a LedgerCryptoProvider or a `YubiCryptoProvider`  as well.

So now we have a way to persist all this info in the form if SecureItems, so next step is to have a 'vault' to store this items:

### `FSSecureStorage` implements `ISecureStorage`
Let's suppose I want to store this items in a local FS. This goes up one level more of abstraction: we split the WHAT we store from the WHERE. Examples here are filesystem, keychain, cloud storage, etc, vaults.

Lastly, since we have the WHAT to build, HOW to identify it, WHICH configs should be applied, HOW it is STORED, and WHERE is is STORED, we're still missing the HOW the CryptoProvider is BUILT:

### `LedgerCryptoProviderBuilder` implements `ICryptoProviderBuilder`
These implementations know how to build the desired provider and how to apply its configs. This is nothing more than a specialized factory pattern for CryptoProviders. Each builder also can create a CryptoProvider from different sources (some could be added/deleted at implementation time): FromSeed, from FromMnemonic, FromSecureItem. The latter will be the most used in runtime since the others are more CLI oriented for UX.

Hope this makes the idea clearer. Happy to give more details if needed.